evm/src/arithmetic/utils.rs

```diff
@@ -17,9 +31,297 @@ pub(crate) fn _range_check_error<const RC_BITS: u32>(file: &str, line: u32, cols
+    b: [ExtensionTarget<D>; N_LIMBS],
+) -> [ExtensionTarget<D>; 3 * N_LIMBS - 1] {
+    let zero = builder.zero_extension();
+    let mut res = [zero; 3 * N_LIMBS - 1];
+    for (i, &ai) in a.iter().enumerate() {
+        for (j, &bj) in b.iter().enumerate() {
+            res[i + j] = builder.mul_add_extension(ai, bj, res[i + j]);
+        }
+    }
+    res
+}
+
+/// Given a(x) and b(x), return a(x)*b(x) mod 2^256.
+pub(crate) fn pol_mul_lo<T, const N: usize>(a: [T; N], b: [T; N]) -> [T; N]
+where
+    T: AddAssign + Copy + Default + Mul<Output = T>,
+{
+    let mut res = pol_zero();
+    for deg in 0..N {
+        // Invariant: i + j = deg
+        for i in 0..=deg {
+            let j = deg - i;
+            res[deg] += a[i] * b[j];
+        }
+    }
+    res
+}
+
+/// Adjoin M - N zeros to a, returning [a[0], a[1], ..., a[N-1], 0, 0, ..., 0].
+pub(crate) fn pol_extend<T, const N: usize, const M: usize>(a: [T; N]) -> [T; M]
+where
+    T: Copy + Default,
+{
+    assert_eq!(M, 2 * N - 1);
+
+    let mut zero_extend = pol_zero();
+    zero_extend[..N].copy_from_slice(&a);
+    zero_extend
+}
+
+pub(crate) fn pol_extend_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    a: [ExtensionTarget<D>; N_LIMBS],
+) -> [ExtensionTarget<D>; 2 * N_LIMBS - 1] {
+    let zero = builder.zero_extension();
+    let mut zero_extend = [zero; 2 * N_LIMBS - 1];
+
+    zero_extend[..N_LIMBS].copy_from_slice(&a);
+    zero_extend
+}
+
+/// Given polynomial a(x) = \sum_{i=0}^{2N-2} a[i] x^i and an element
+/// `root`, return b = (x - root) * a(x).
+///
+/// NB: Ignores element a[2 * N_LIMBS - 1], treating it as if it's 0.
+pub(crate) fn pol_adjoin_root<T, U>(a: [T; 2 * N_LIMBS], root: U) -> [T; 2 * N_LIMBS]
+where
+    T: Add<Output = T> + Copy + Default + Mul<Output = T> + Sub<Output = T>,
+    U: Copy + Mul<T, Output = T> + Neg<Output = U>,
+{
+    // \sum_i res[i] x^i = (x - root) \sum_i a[i] x^i. Comparing
+    // coefficients, res[0] = -root*a[0] and
+    // res[i] = a[i-1] - root * a[i]
+
+    let mut res = [T::default(); 2 * N_LIMBS];
+    res[0] = -root * a[0];
+    for deg in 1..(2 * N_LIMBS - 1) {
+        res[deg] = a[deg - 1] - (root * a[deg]);
+    }
+    // NB: We assume that a[2 * N_LIMBS - 1] = 0, so the last
+    // iteration has no "* root" term.
+    res[2 * N_LIMBS - 1] = a[2 * N_LIMBS - 2];
+    res
+}
+
+pub(crate) fn pol_adjoin_root_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    a: [ExtensionTarget<D>; 2 * N_LIMBS],
+    root: ExtensionTarget<D>,
+) -> [ExtensionTarget<D>; 2 * N_LIMBS] {
+    let zero = builder.zero_extension();
+    let mut res = [zero; 2 * N_LIMBS];
+    let t = builder.mul_extension(root, a[0]);
+    res[0] = builder.sub_extension(zero, t);
+    for deg in 1..(2 * N_LIMBS - 1) {
+        let t = builder.mul_extension(root, a[deg]);
+        res[deg] = builder.sub_extension(a[deg - 1], t);
```

Comment:
This could be done with one `arithmetic_extension`. Something like
```suggestion
        res[deg] = builder.arithmetic_extension(F::NEG_ONE, F::ONE, root, a[deg], t);
```